desktop/core/src/desktop/lib/botserver/views_tests.py

```diff
@@ -75,34 +75,66 @@ def test_handle_on_message(self):
+            with patch('desktop.lib.botserver.views.slack_email_prefix') as email_prefix:
 
+              # Slack user email: test@example.com
               client = make_logged_in_client(username="test", groupname="default", recreate=True, is_superuser=False)
               user = User.objects.get(username="test")
+              # Other slack user email: test_not_me@example.com
+              client_not_me = make_logged_in_client(username="test_not_me", groupname="default", recreate=True, is_superuser=False)
+              user_not_me = User.objects.get(username="test_not_me")
+
               channel_id = "channel"
               message_ts = "12.1"
+              user_id = "<@user_id>"
 
-              # qhistory link
-              links = [{"url": "https://demo.gethue.com/hue/editor?editor=123456"}]
+              # Query link
+              links = [{"url": "https://demo.gethue.com/hue/editor?editor=12345"}]
               doc_data = {
                 "dialect": "mysql",
                 "snippets": [{
                   "database": "hue",
                   "statement_raw": "SELECT 5000",
                 }]
               }
-              document2_objects_get.return_value = Mock(data=json.dumps(doc_data), owner=user)
+              doc = Document2.objects.create(id=12345, data=json.dumps(doc_data), owner=user)
               mock_unfurl_payload.return_value = {
                 'payload': {},
                 'file_status': True,
               }
-              handle_on_link_shared(channel_id, message_ts, links)
+
+              # Other user sends link to unfurl without read permission
+              email_prefix.return_value = 'test_not_me'
+
+              assert_raises(PopupException, handle_on_link_shared, "channel", "12.1", links, "<@user_id>")
+              assert_false(chat_unfurl.called)
+              assert_false(send_result_file.called)
+
+              # User doesn't exist and read perm is also not set for link
+              email_prefix.return_value = 'test_user_not_exist'
+
+              assert_raises(PopupException, handle_on_link_shared, "channel", "12.1", links, "<@user_id>")
+              assert_false(chat_unfurl.called)
+              assert_false(send_result_file.called)
+
+              # Other user sends link which has read perm
+              email_prefix.return_value = 'test_not_me'
+              doc.share_link(user, perm='read') # Some user has set read perm to shared link
+
+              handle_on_link_shared(channel_id, message_ts, links, user_id)
               assert_true(chat_unfurl.called)
               assert_true(send_result_file.called)
 
-              # gist link
+              # Doc owner sends link to unfurl
+              email_prefix.return_value = 'test'
+              handle_on_link_shared(channel_id, message_ts, links, user_id)
+
+              assert_true(chat_unfurl.called)
+              assert_true(send_result_file.called)
+             
+              # Gist link
```

Comment:
Also probably time to split the test function in parts testing each one function/flow

Reply:
Split in different link share flows